File: rail-fence/Python/decipher.py
from math import floor
import sys

def decipher(cipher_text: str, depth: int):
    ct_len = len(cipher_text)
    tick_size = (2 * depth) - 2
    p = floor( ct_len / tick_size )
    r = ct_len % tick_size
    
    element_count = [0 for _ in range(depth)] # stores no of elements in each row
    
    element_count[0] = p
    element_count[-1] = p

    if r > depth:
        element_count[0] += 1
        element_count[-1] += 1

    # processing middle rows
    i = 1
    while i < len(element_count) - 1:
        element_count[i] = p * 2
        i += 1

    if r > depth:
        i = depth - 2
        while i > r - depth:
            element_count[i] += 1
            i -= 1
    i = 0
    while i < min(r, depth):
        element_count[i] += 1
        i += 1

    indices = [0]
    total = 0
    for i in element_count[:-1]:
        indices.append(total + i)
        total += i

    i = 0
    pt = []
    direction = 1

    while len(pt) < len(cipher_text):
        index = indices[i]
        pt.append(cipher_text[index])
        indices[i] += 1

        if i == 0: direction = 1
        elif i == len(indices) - 1: direction = -1

        i += direction
    
    return "".join(pt)

if __name__ == "__main__":
    print(decipher(sys.argv[1], int(sys.argv[2])))

File: rail-fence/C/decipher.c
#include <stdio.h>
#include <string.h>
#include <math.h>
#include "linked_list.h"

char *decipher(char *cipher_text, int depth) {
    int ct_len = strlen(cipher_text);
    int tick_size = (2 * depth) - 2;
    int p = floor(ct_len / tick_size);
    int r = ct_len % tick_size;

    int element_count[depth];
    element_count[0] = p;
    element_count[depth - 1] = p; // last element

    if(r > depth) {
        element_count[0]++;
        element_count[depth-1]++;
    }

    for(int i = 1; i < depth - 1; i++) {
        element_count[i] = p * 2;
    }

    if(r > depth) {
        int i = depth - 2;
        while(i > r - depth) {
            element_count[i]++;
            i--;
        }
    }

    int i = 0;
    while(i < (r < depth ? r : depth)) {
        element_count[i]++;
        i++;
    }

    int indices[depth];
    indices[0] = 0;
    int total = 0;
    for(int i = 0; i < depth - 1; i++) {
        indices[i+1] = total + element_count[i];
        total += element_count[i];
    }

    char *pt = (char*)malloc(sizeof(char) * ct_len);
    i = 0;
    int direction = 1;

    int pt_index = 0;
    while(pt_index < ct_len) {
        int index = indices[i];
        pt[pt_index++] = cipher_text[index];
        indices[i]++;

        if(i == 0)
            direction = 1;
        else if(i == depth - 1)
            direction = -1;
        i += direction;
    }

    pt[pt_index] = '\0';

    return pt;
}

int main(int argc, char **argv) {
    printf("%s\n", decipher(argv[1], atoi(argv[2])));
}